stdlib/public/core/String.swift

```diff
@@ -925,91 +925,57 @@ extension String {
 }
 
 // Index conversions
-extension String.Index {
-  /// Construct the position in `characters` that corresponds exactly to
-  /// `unicodeScalarIndex`. If no such position exists, the result is `nil`.
-  ///
-  /// - Precondition: `unicodeScalarIndex` is an element of
-  ///   `characters.unicodeScalars.indices`.
-  public init?(
-    _ unicodeScalarIndex: String.UnicodeScalarIndex,
-    within characters: String
-  ) {
-    if !unicodeScalarIndex._isOnGraphemeClusterBoundary {
-      return nil
-    }
-    self.init(_base: unicodeScalarIndex)
-  }
-
-  /// Construct the position in `characters` that corresponds exactly to
-  /// `utf16Index`. If no such position exists, the result is `nil`.
-  ///
-  /// - Precondition: `utf16Index` is an element of
-  ///   `characters.utf16.indices`.
-  public init?(
-    _ utf16Index: String.UTF16Index,
-    within characters: String
-  ) {
-    if let me = utf16Index.samePosition(
-      in: characters.unicodeScalars
-    )?.samePosition(in: characters) {
-      self = me
-    }
-    else {
-      return nil
-    }
-  }
-
-  /// Construct the position in `characters` that corresponds exactly to
-  /// `utf8Index`. If no such position exists, the result is `nil`.
-  ///
-  /// - Precondition: `utf8Index` is an element of
-  ///   `characters.utf8.indices`.
-  public init?(
-    _ utf8Index: String.UTF8Index,
-    within characters: String
-  ) {
-    if let me = utf8Index.samePosition(
-      in: characters.unicodeScalars
-    )?.samePosition(in: characters) {
-      self = me
-    }
-    else {
-      return nil
-    }
-  }
-
-  /// Returns the position in `utf8` that corresponds exactly
-  /// to `self`.
-  ///
-  /// - Precondition: `self` is an element of `String(utf8).indices`.
+extension String {
+  // TODO: swift-3-indexing-model - add docs
   @warn_unused_result
-  public func samePosition(
-    in utf8: String.UTF8View
-  ) -> String.UTF8View.Index {
-    return String.UTF8View.Index(self, within: utf8)
+  public func convert(
+    index index: String.UTF8Index,
+    from view: String.UTF8View
+  ) -> String.Index? {
```

Comment:
The old API lived on the String.XxxIndex and you asked that index to give you an index of a different type related to the supplied view. In the old model the Indexes had references to the same string core. In the new model Indexes don't have (or shouldn't have to have) a reference to the string core.

To support uniformity of going back to your collection to manipulate an Index (next(i), advance(i, by:10), etc.) I moved the index transforms to their related views (e.g. collections). I think that requires an API naming change away from "samePosition" since you have to supply the index to transform.

In the end it came down to something like the following (at least in my mind)
```
extension String {
  public func convert(index i: String.UTF8Index, from view: String.UTF8View) -> String.Index?
}
```
...or...
```
extension String.UTF8View {
  public func convert(index i: String.UTF8Index, to view: String) -> String.Index?
}
```
I went for the former (not sure why :smile: now) but now that I stepped away from it for a bit I think the the later is more easily understood.